src/collects/seashell/backend/lock.rkt

```diff
@@ -17,34 +17,104 @@
 ;; You should have received a copy of the GNU General Public License
 ;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-(provide acquire-exclusive-write-lock
-         release-exclusive-write-lock
-         call-with-exclusive-write-lock
-         acquire-write-lock
-         release-write-lock
+(require/typed ffi/unsafe/atomic
+               [start-atomic (-> Void)]
+               [end-atomic (-> Void)])
+
+(provide call-with-exclusive-write-lock
          call-with-write-lock)
 
 ;; Some arbitrary number of concurrent writes to block at.
 ;; We should probably allow a couple of concurrent writes if students have
 ;; different windows open.. more than a few wouldn't be useful though
 (define concurrent-writes 10)
-(define write-sem (make-semaphore concurrent-writes))
+
+;; Two custom synchronization primitves are defined below.
+;; The safe-lock primitive is a general-purpose lock which detects dead threads holding locks.
+;; The safe-semaphore is not as general. It enforces that you must wait for a semaphore
+;; before posting to it. This is to easily allow checking for unbalances caused by dead
+;; threads. Keep this in mind before using it for something new.
+(struct safe-semaphore ([count : Integer] [threads : (Listof Thread)]) #:mutable)
+(struct safe-lock ([thread : (U Thread False)]) #:mutable)
+
+(: make-safe-semaphore (-> Integer safe-semaphore))
+(define (make-safe-semaphore count)
+  (safe-semaphore count '()))
+
+(: safe-semaphore-post (-> safe-semaphore Void))
+(define (safe-semaphore-post sem)
+  (start-atomic)
+  (cond
+    [(member (current-thread) (safe-semaphore-threads sem))
+     (set-safe-semaphore-count! sem (+ 1 (safe-semaphore-count sem)))
+     (set-safe-semaphore-threads! sem (remove (current-thread) (safe-semaphore-threads sem)))]
+    [else (end-atomic) (error "Unbalanced safe-semaphore post before wait")])
+  (end-atomic))
+
+(: safe-semaphore-wait (-> safe-semaphore Void))
+(define (safe-semaphore-wait sem)
+  (start-atomic)
+  (define dead-thds (filter thread-dead? (safe-semaphore-threads sem)))
+  (set-safe-semaphore-count! sem (- (safe-semaphore-count sem) (length dead-thds)))
+  (set-safe-semaphore-threads! sem (remove* dead-thds (safe-semaphore-threads sem)))
+  (cond
+    [(< 0 (safe-semaphore-count sem))
+     (set-safe-semaphore-count! sem (- (safe-semaphore-count sem) 1))
+     (set-safe-semaphore-threads! sem (cons (current-thread) (safe-semaphore-threads sem)))
+     (end-atomic)]
+    [else (end-atomic) (sleep 0) (safe-semaphore-wait sem)]))
```

Comment:
This is going to eat CPU time spinning/waiting on the lock.  You probably ought to  have a list of threads that are suspended on the lock and for safe-semaphore-post to randomly wake up a thread that is suspended.